### PR TITLE
Add Crate Owner Auditing

### LINKED ISSUE #1548: Audit trail for more actions
Further steps beyond #1478. These changes wouldn't be exposed in the UI but would be in the database for us to use in determining who took what actions when.

- [ ] Add a column to the api_tokens table named `revoked` that's a BOOLEAN NOT NULL DEFAULT 'f'. Then instead of deleting tokens, mark them as revoked.

- [ ] Create a migration (using `diesel migration generate`, [as explained in this guide](http://diesel.rs/guides/getting-started/)) that adds a new table named `version_owner_actions` with the following columns:
  - id SERIAL PRIMARY KEY
  - version_id FOREIGN KEY REFERENCES versions (id)
  - owner_id FOREIGN KEY REFERENCES users (id)
  - owner_token_id FOREIGN KEY REFERENCES api_tokens (id)
  - action INTEGER NOT NULL (that would map to a Rust enum with variants publish, yank, unyank)
  - time TIMESTAMP NOT NULL DEFAULT now()

- [ ] Migrate the current `published_by` column on the `versions` table to be actions in the `version_owner_actions` table instead.

- [ ] Keep the [headers[0]](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/middleware/current_user.rs#L45) authorization header value accessible by holding onto it in the [`AuthenticationSource::ApiToken` variant](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/middleware/current_user.rs#L18)

- [ ] Within the [crate publish transaction](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/controllers/krate/publish.rs#L69), [after the new version record is created](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/controllers/krate/publish.rs#L129-L139), create a new version owner action record with action = "publish", version_id = [version.id](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/controllers/krate/publish.rs#L129-L139), owner_id = [user.id](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/controllers/krate/publish.rs#L42), owner_token = req.authentication_source() (and then extract the 
token value)

- [ ] Make similar changes to add records to the activity table in [yank and unyank](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/controllers/version/yank.rs#L24)

- [ ] Add a field to [EncodableVersion](https://github.com/rust-lang/crates.io/blob/edcde838405303cc1afe5a0e6ff895a69fef96fb/src/views/mod.rs#L185) that's a Vec of all the actions, owner ids, and times that this version has had an action taken on it so that this info, minus the api token value, is returned in the API response

- [ ] Add a table crate_owners_actions that records who adds and removes other owners from a crate

- [ ] Add tests that this information is being recorded and returned as expected

### PR DESCRIPTION
**Pull request by markcatley**
Including the following audit actions:
- Invite User
- Remove User

Ticks off one of the items in #1548.

### CODE REVIEW

**Comment by smarnach** [comment]
Thanks for working on this! Unless someone else picks this up before then, I will hopefully get to this on 27 December.

File: src/views.rs [hunk 1]
```diff
@@ -103,6 +103,7 @@ pub struct EncodableCrate {
     pub repository: Option<String>,
     pub links: EncodableCrateLinks,
     pub exact_match: bool,
+    pub audit_actions: Vec<EncodableAuditAction>,
```
Comment (smarnach) [inline src/views.rs]: This publicly exposes the audit information via the API, right? We need to discuss whether this should be public. In any case, if this is available via the API, it should also be included in the database dumps and vice versa. Currently this PR excludes the information from the DB dumps.

File: src/models/krate_owner_actions.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+};
+use std::io::Write;
+
+use crate::models::{ApiToken, Crate, User};
+use crate::schema::*;
+
+#[derive(Debug, Clone, Copy, PartialEq, FromSqlRow, AsExpression)]
+#[repr(i32)]
+#[sql_type = "Integer"]
+pub enum CrateAction {
```
Comment (smarnach) [inline src/models/krate_owner_actions.rs]: I'd prefer a name like `ActionKind` or `ActionType` here, or even `CrateOwnerActionKind`, though this is kind of unwieldy.

We have the `CrateOwnerAction` and `CrateAction` types in this module, and while these types have quite different nature, the nature of their difference isn't reflected in the current naming.

**Comment by markcatley** [comment]
Hi, thanks for the review.

This mostly follows the instructions in the issue, perhaps @carols10cents can weigh in on the rational. I will try to answer the specific questions where I can. I will note that this follows the same pattern as version owner actions that was done as part of the same issue.

_Do we want the audit information to be public?_
That's an interesting question. I cannot think of any strong arguments either way. What's the default answer to this question in the project? I think the argument is stronger for Version Owner Actions being public as I can see it being useful knowing who published/yanked/unyanked a crate and when. In absence of a strong argument I would probably advocate following the lead of that change.

_Do we also want to track when invitations are accepted or rejected?_
I think that's a good idea. I'll look at adding it when I am back from holiday next week.

_The auditing information is already partially available in the crate_owners and crate_owner_invitation tables. We should consisder deduplicating this._
This is an interesting idea, I could create an issue and do it as a follow up if the information is available to backfill the information. I looked at it for version owner actions and I didn't think the information was complete enough.

File: src/models/krate_owner_actions.rs (hunk 2 above)
Comment (markcatley) [reply src/models/krate_owner_actions.rs]: I like the suggestion. I'll have a look.

File: src/views.rs (hunk 1 above)
Comment (markcatley) [reply src/views.rs]: Once we resolve the issue I'll update this, and for version owner actions. I wasn't that familiar with what's expected here.

**Comment by carols10cents** [comment]
I was imagining information like "user x added user y as an owner at datetime" being available publicly so that anyone could audit the risk of activities being malicious or not. One use case this could enable is what dependabot does for JavaScript: for example, [in this PR](https://github.com/rust-lang/crates.io/pull/2033), dependabot has highlighted maintainer changes: 

> Maintainer changes
> <p>This version was pushed to npm by <a href="https://www.npmjs.com/~eslintbot">eslintbot</a>, a new releaser for eslint since your current version.</p>

Does that help?

**Comment by carols10cents** [comment]
> _Do we also want to track when invitations are accepted or rejected?_
> I think that's a good idea. I'll look at adding it when I am back from holiday next week.
> 

I feel like knowing when an invitation is sent and when an invitation is rejected are not useful pieces of information; it's only when an invitation is accepted and a crate has a new owner that it's interesting. What do you think?

**Comment by smarnach** [comment]
Thanks for the input, @carols10cents!

I agree that it makes sense to make changes in the ownership public, so let's make the new table public.

If that information is public, we should _exclude_ information about when invitations are created and rejected, since this information should not be public.

The `created_at`, `created_by` and `updated_at` columns (and the corresponding triggers) can be removed from the `crate_owners` table, since we will have a better source for that information. The `crate_owner_invitations` table can remain unchanged. The new auditing table should be backfilled with the information from the deleted columns. This can either happen in a follow-up PR or in this one.

I think that settles all open questions. @markcatley Does this sound good to you?

**Comment by locks** [comment]
Hi @markcatley! Are you able to address the latest discussion points?

**Comment by Turbo87** [comment]
closing due to inactivity

**Comment by carols10cents** [comment]
I know this is closed now, and I'll add this on the related issue too, but in thinking about this more, we do need to think about cases when a person might want to remove all association they have with a crate, past and present. So user_a, who was owner of crate_b at some point, now wants nothing to do with crate_b and doesn't want crates.io to show that they were ever owner of crate_b.

**Pull request closed without merging**